allensdk/api/cloud_cache/cloud_cache.py

```diff
@@ -39,9 +46,90 @@ def __init__(self, cache_dir, project_name):
 
         self._manifest = None
         self._cache_dir = cache_dir
+
         self._project_name = project_name
         self._manifest_file_names = self._list_all_manifests()
 
+        # what latest_manifest was the last time an OutdatedManifestWarning
+        # was emitted
+        self._manifest_last_warned_on = None
+
+        # self._downloaded_data_path is where we will keep a JSONized
+        # dict mapping paths to downloaded files to their file_hashes;
+        # this will be used when determining if a downloaded file
+        # can instead be a symlink
+        c_path = pathlib.Path(self._cache_dir)
+        self._downloaded_data_path = c_path / '_downloaded_data.json'
+
+        if not self._downloaded_data_path.exists():
+            self._construct_file_version_lookup()
+
+    def _construct_file_version_lookup(self) -> None:
```

Comment:
is this potentially a 3-hour long loop?
I thought hashing all the NWBs for upload took that long. So, if a user happened to have already downloaded all NWBs, would it take that long?

Reply:
Yeah. It won't be pretty.

This loop will only run when the user instantiates a cache to a directory that does not contain `self._downloaded_data_paths`. That can either happen because

a) a pre existing cache got corrupted somehow (i.e. `self._downloaded_data_path` got deleted)
b) a user who downloaded NWB files before this PR got merged updates the code and resumes using an old cache

In a perfect world, `self._downloaded_data_path` will get saved in the cache directory and this method will never run.

The concern was that users in category (b) will not benefit from the functionality in this PR, and will end up downloading duplicate copies of NWB files when they download the updated data files. Either way, a time consuming process is involved (downloading duplicate NWB files or building `self._downloaded_data_path` in an already existing cache).

Possible solutions that come to my mind

1) just don't build the local manifest from scratch. Users in case (b) will have to eat the cost of downloading duplicate NWBs once when this PR gets merged

2) Add a warning to this method alerting the user that "the local data manifest is missing; I am going to try to rebuild it; this could be time consuming; if you don't want to do that, abort and re-initialize your cache with ...." and then add a kwarg to `__init__.py` that disables this method.

3) Make this method user-facing. If the `CloudCache` finds that there are NWB files in cache_dir, but no `self._downloaded_data_path` file, it can emit a warning encouraging the user to run `construct_file_version_lookup` by hand, warning them that it may be time consuming.

In either (2) or (3), I can try to add a status bar to `construct_file_version_lookup`.

I think (3) might be the way to go.

Reply:
I've changed the logic so that, if on instantiation the constructor thinks that the local manifest was accidentally deleted, it emits a warning suggesting that the user run `construct_local_manifest`.

There is now also a status bar in `construct_local_manifest` so users can see the progress in file hashing.